### PR TITLE
GZipUncompressorOutputStream + and more

### PR DESCRIPTION
**Pull request by DimaGolomozy**
1. improve GZipUncompressorOutputStream (the code is taken from GzipInputStream):
- removed piping output/input stream
- removed piping thread

2. removed `DecodingOutputStream` in favor of `UnCloseableOutputStream`,
the `DecodingOutputStream` was there only for protecting the "clients" outputstream from closing, its simpler to wrap it with a wrapper that just wont close the clients outputstream. this also fixes a "bug" in https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/DeflateDecoder.java#L17
where in `InflaterOutputStream` is not closed.

3. exposing more methods to control buffer size from the client.

 @renatoathaydes ^

### CODE REVIEW

**Comment by renatoathaydes** [comment]
@DimaGolomozy I don't see what problem this PR is trying to solve. What is the problem with the current implementation of the gzip decoder that you're addressing? Using code internal to `GZIPInputStream` is not an option for this library as we don't want to focus on the gzip algorithm, we want to focus on HTTP only. I don't have the knowledge to maintain the code you extracted from `GZIPInputStream` and unless there is a very strong reason to do so, I would prefer to keep using that class as it is made available by its API (which unfortunately requires a Thread).

File: rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java [hunk 1]
```diff
@@ -47,29 +47,27 @@ public BodyDecoder(HttpBodyEncodingRegistry registry, List<String> transferEncod
...
-     * Create a {@link DecodingOutputStream} that decodes the contents written to it before passing it on
+     * Create a decoder {@link OutputStream} that decodes the contents written to it before passing it on
      * to the given {@link OutputStream}.
      *
      * @param out receiver of decoded contents
      * @return a stream that decodes the bytes written into it, then writes the decoded bytes into another stream
      * @throws IOException if an error occurs while writing to the given stream
      */
-    public DecodingOutputStream decoding(OutputStream out) throws IOException {
+    public OutputStream decoding(OutputStream out) throws IOException {
```
Comment (renatoathaydes) [inline rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java:57]: Isn't this a pointless API change? Why return a less specific type?

Comment (DimaGolomozy) [reply rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java:57]: the `DecodingOutputStream` is very limiting because it extends `FilterOutputStream` what if i want to add a new XOutputStream that uses `InflaterOutputStream` you wont be able to do that. (unless you add the `InflaterOutputStream` as a variable in the class, but then you lose the point of FilterOutputStream)

plus, as i wrote, the `DecodingOutputStream` is used only for "not closing the clients output stream" (or am i wrong?)
the `finishDecoding()` is only used just the close.
so why not create a OutputStream that just dont closes the clients output stream?  see `UnCloseableOutputStream`
this also fixes a scenario in:
https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/DeflateDecoder.java#L17
where in `InflaterOutputStream` is never closed

**Comment by DimaGolomozy** [comment]
> @DimaGolomozy I don't see what problem this PR is trying to solve. What is the problem with the current implementation of the gzip decoder that you're addressing? Using code internal to `GZIPInputStream` is not an option for this library as we don't want to focus on the gzip algorithm, we want to focus on HTTP only. I don't have the knowledge to maintain the code you extracted from `GZIPInputStream` and unless there is a very strong reason to do so, I would prefer to keep using that class as it is made available by its API (which unfortunately requires a Thread).

1. performance;
on each request with body, creating an `executorService` 
https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java#L35
and starting a thread that will just copy the data from 1 stream to the other

2. no need blocking and waiting for the execution (also impacts performance):
https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java#L89
its very optimistic to give 5 seconds :D
but what happen if it will take 6 seconds to finish piping the data between the input streams?

3.  remove the need to propagate exceptions/interrupt between the threads;
https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/GZipUncompressorOutputStream.java#L70-L73

basically removes the need to the thread that just pipes the data between the streams, and no blocking.

**Comment by renatoathaydes** [comment]
So, your reason is basically performance... but how much faster is the new approach for realistic client/server workloads? Are you sure that the IO cost doesn't completely shadows the cost of maintaining a decoder thread? IO seems orders of magnitude slower to me, so I don't share the same confidence performance will be improved.

>  but what happen if it will take 6 seconds to finish piping the data between the input streams?

It's 5 seconds from the last time the writer called `write` followed by `close`. Because `write` is a blocking call, even on the piped stream, this means 5 seconds from the last write, which is pretty much guaranteed to be enough.

File: rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java (hunk 1 above)
Comment (renatoathaydes) [reply rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java:57]: > the DecodingOutputStream is very limiting because it extends FilterOutputStream what if i want to add a new XOutputStream that uses InflaterOutputStream

You mean this is limiting for the implementator of `HttpMessageDecoder`? I disagree.

Besides making it clear the semantics of the decoder stream being returned, it allows an implementation to explicitly run some code at the end, which is very common for decoders to do as they may need to know when the data is fully consumed.

I wanted to make it very clear that the implementations of `DecodingOutputStream` should not close the underlying stream, while still having a clear hook for implementing cleanup/finalizing the decoding algorithm. This is not a limitation, this is a useful restriction to behaviour implementations are expected to follow.

I see no reason to remove this type and I won't accept any PRs removing it, sorry.

>  where in InflaterOutputStream is never closed

We cannot call `close` on that as it would close the underlying stream. You're probably right that we should be wrapping the outputStream given to it to avoid this problem, please make a separate ticket for that if you think this is a problem (we're flushing the stream, but I'm not sure that's enough).

**Comment by renatoathaydes** [comment]
This PR cannot be accepted as it is because:

* it changes the public API of the library.
* it introduces zip-stream logic into it (copied from `GZIPInputStream`), which is not something I am prepared to maintain in this library, which is HTTP-focused.

I would be very happy to accept a changeset that made the GZIP decoder simpler and avoided a Thread, though, as long as it did not require knowledge of GZIP internals to be transferred to the code implemented in this library.

Regarding performance: if you can show that maintaining a thread for decoding zip streams is a completely unnacceptable CPU performance overhead in cases of clients/servers operating in realistic scenarios (i.e. using real networking, not just local sockets) I might reconsider this decision. Thread overhead in terms of memory is not an acceptable argument because this library is already limited to synchronous IO (requiring one thread per simultaneous connection), and the fact that Project Loom will, in due time, make JVM Threads cheaper enough for this to become a non-issue.
